Monorepo Utils: Add changelog addition from actions

psealock opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Adds an action to call the [CLI changefile addition tool](https://github.com/woocommerce/woocommerce/pull/38357).

The CLI will be called on pull request creation, added commits, and description editing. A template edition that adds fields for authors to fill out is also included. When the box is ticked, the automation will add changefile entries to all affected projects based on the data supplied in the description.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Fork this repo and push this branch (`add/add-changelog-action`) to your fork.
2. Merge this branch into `trunk`.
3. Create a new branch from `trunk` called `test/add/add-changelog-action` and make a few changes to projects with no changelogger, ie `tools/`. Commit and push changes to your fork. Create a pull request into your fork's `trunk` branch. The pull request will now include a section to enter changelog details. Leave this section unchanged.
4. See the `Changelog Auto Add` PR check kick off. Click to view. Confirm the job succeeds and results in a message similar to `PR #71 does not have the "Automatically create a changelog entry from the details" checkbox checked. No changelog will be created.`
5. Now edit the PR description and fill out the changelog details. The `Changelog Auto Add` will kick off again once you save the description. See the check succeed once again with a message like `No projects require a changelog`.
6. Now make a change in a project that does have changelogger, like `woocommerce-admin`. Once changes are committed and pushed up, see the `Changelog Auto Add` check kickoff once again. This time, a changefile will be added based on the data provided in the PR description.
7. Confirm the changefile create exists and reflects the data in the description.
8. Change the description, save, and see the job run again. All changefiles will reflect the change.
9. Now undo a change in a project. See the changefile removed.


<!-- End testing instructions -->

psealock committed:
add basic CLI command

psealock committed:
enable on PRs

psealock committed:
add prNumber

psealock committed:
clone full repo

psealock committed:
avoid endless loop

psealock committed:
clean up code

psealock committed:
better message

samueljseay commented on bin/pre-push.sh:
```diff
@@ -31,9 +31,3 @@ if [ $? -ne 0 ]; then
-git -c core.hooksPath=/dev/null checkout $PROTECTED_BRANCH --quiet
-git -c core.hooksPath=/dev/null checkout $CURRENT_BRANCH --quiet
-
-php tools/monorepo/check-changelogger-use.php $PROTECTED_BRANCH $CURRENT_BRANCH
```
👏🏻 👏🏻 👏🏻 👏🏻

samueljseay commented on tools/monorepo-utils/src/changefile/lib/projects.ts:
```diff
@@ -72,27 +77,32 @@ export const getChangeloggerProjectPaths = async (
 
-	// make sure base sha is available.
+	//make sure base sha is available.
 	await git.raw( [
```
Nitpick there is `git.addRemote` for this.

psealock replied:
fixed in https://github.com/woocommerce/woocommerce/pull/38510/commits/bc38e72a64070f1cc95c37e492f8dcb06a1d9ede

samueljseay commented on tools/monorepo-utils/src/changefile/lib/projects.ts:
```diff
@@ -72,27 +77,32 @@ export const getChangeloggerProjectPaths = async (
+		getAuthenticatedRemote( { owner: baseOwner, name: baseName } ),
 	] );
-	await git.raw( [ 'fetch', 'woocommerce', base ] );
+	await git.raw( [ 'fetch', baseOwner, base ] );
```
There is `git.fetch` for this

psealock replied:
fixed in https://github.com/woocommerce/woocommerce/pull/38510/commits/bc38e72a64070f1cc95c37e492f8dcb06a1d9ede

samueljseay commented on tools/monorepo-utils/src/core/git.ts:
```diff
@@ -115,6 +115,24 @@ export const cloneRepoShallow = async ( repoPath: string ) => {
+} ) => {
+	const { owner, name } = options;
+	const source = `github.com/${ owner }/${ name }`;
+	const token = getEnvVar( 'GITHUB_TOKEN' );
```
Just a question, do we always want to pass `true` to this I think? to ensure it errors if it is not present?

psealock replied:
Yes, I think so. Fixed in https://github.com/woocommerce/woocommerce/pull/38510/commits/2072f74190d1a7bf74a3e83b810ec1245bf3c5e7

samueljseay commented on .github/workflows/changelog-auto-add.yml:
```diff
@@ -1,10 +1,48 @@
+            - name: Checkout code
+              uses: actions/checkout@v3
+              with:
+                  fetch-depth: 0
```
This is intentional ay because you want all the commits from the whole PR?

Not necessary change to make but for future it would be cool to explore just fetching commits for the PR, cause this fetches all branches and tags.

I saw something like this with a quick google:

```
- name: 'PR commits + 1'
  run: echo "PR_FETCH_DEPTH=$(( ${{ github.event.pull_request.commits }} + 1 ))" >> "${GITHUB_ENV}"

- name: 'Checkout PR branch and all PR commits'
  uses: actions/checkout@v3
    with:
      ref: ${{ github.event.pull_request.head.ref }}
      fetch-depth: ${{ env.PR_FETCH_DEPTH }}
 ```

psealock replied:
>This is intentional ay because you want all the commits from the whole PR?

No, good catch. When the CLI tool clones the repo, yes there needs to be a commit history for calculation the diff but not on this initial checkout. Fixed in https://github.com/woocommerce/woocommerce/pull/38510/commits/047ff513e38744b8fc352e6a58f6a11d21b12a81

samueljseay commented on tools/monorepo-utils/src/changefile/index.ts:
```diff
@@ -133,6 +128,22 @@ const program = new Command( 'changefile' )
+					Logger.notice(
+						`Installing dependencies in ${ tmpRepoPath }`
+					);
+					execSync( 'pnpm install', {
```
Wow so we call the composer changelogger directly on each package. Is that why we do this? I see the need of course but, this is rough 😂 

Its fine for now obviously, but question, can the changelogger be installed at the top level of the repo and just run for each package path in future or something? Just a thought, otherwise I guess I'd really like to see us make that port of the changelogger to JS.

psealock replied:
>Wow so we call the composer changelogger directly on each package. Is that why we do this? I see the need of course but, this is rough 😂

yes, its rough. And the reason why I put this step until the last possible. I tried doing just `composer install`, but there are some JS dependencies too for some packages.

>can the changelogger be installed at the top level of the repo and just run for each package path in future or something?

Thats a good idea. I will look into it as a follow up

samueljseay reviewed (changes requested):
So some nitpicks and questions from me. Those are not blockers, up to you if you handle them 😄 

but one thing about the UX I think we need to tweak:

If for some reason you leave out a value or your message is not properly formed it causes the changelogger to fail. I think in the case the user checks the changelog box, we need to validate the inputs first and if they're wrong the action should not fail with error, but instead post a comment to the PR telling them what's invalid about the changelog they tried to enter?

It could also fail as well, but it should output some errors that make sense to a user I think? For example right now it just fails with changelogger error `Entry must be specified in non-interactive mode.`, which I understand what's gone wrong, but for presenting this as UX for PR I think we need it to be a bit more helpful with validation?

samueljseay commented:
Oh also I wondered what about enclosing the message and comment in backticks instead of the HTML comment?

At first I wasn't sure if the template expected me to delete the comments or not. if its in backticks it's immediately visible in both edit and preview? Not married to the idea it just felt more intuitive to me.

psealock committed:
add validation

psealock committed:
lock file

psealock committed:
use git fns

psealock committed:
pass true to getEnvVar

psealock committed:
fetch just 1 commit

psealock committed:
lock file update

psealock committed:
lock file update

psealock commented:
Note to self: `isCommunityPullRequest` should not use `author_association` because it can be changed by the author.

psealock committed:
remove use of author_association to determince communityPRs

samueljseay commented on tools/monorepo-utils/src/changefile/lib/__tests__/github.ts:
```diff
@@ -383,19 +382,73 @@ describe( 'getChangelogDetails', () => {
 		);
 	} );
+
+	it( 'should return an error when no sifnificance found', () => {
```
```suggestion
	it( 'should return an error when no significance found', () => {
```

samueljseay commented on tools/monorepo-utils/src/changefile/lib/github.ts:
```diff
@@ -130,38 +130,62 @@ export const getChangelogMessage = ( body: string ) => {
-		message,
-		comment,
-	};
+	if ( ! significance ) {
```
Remember if you're checking for the presence of something you're implying it could not be there, that means the type for `significance` should not be `significance: string` it should be `significance?: string`

this applies to all the other arguments you're checking for existence of here.

samueljseay commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -28,3 +28,37 @@ Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerc
+-   [ ] Performance - Address performance issues
+-   [ ] Enhancement
+
+#### Message <!-- Add a changelog message here -->
```
You want the message to appear on newline after the "Message" heading right? I found this really confusing when trying to put the message in the right place and it fails if message is on same line as heading. I still argue this needs to be clearer somehow. Perhaps the regex needs to be more flexible, swallowing all leading whitespace after the "Message" heading and trailing whitespace up until the "Comment" heading.

samueljseay reviewed (commented):
Some nitpicks and suggestions from me, but the main issue I'm having is that the action now seems unable to pick up the message I added to my PR description: https://github.com/samueljseay/woocommerce/pull/61

It still feels really unintuitive not knowing why its failing, I wonder if you need to make the regex much more flexible. In comments on review I've suggested matching/swallowing whitespace up until you find a non-whitespace character rather than what looks like a very restricted regex you have right now. It could be naive though because I haven't tested your regex, since I can't seem to get it working right now, I hope my PR example can help you debug further.

psealock committed:
Update tools/monorepo-utils/src/changefile/lib/__tests__/github.ts

Co-authored-by: Sam Seay <samueljseay@gmail.com>

psealock committed:
fix error logic

psealock committed:
fix regex

psealock committed:
designate optional params in getChangelogDetailsError

psealock committed:
apply regex fix to comments too

psealock commented:
Thanks for the review @samueljseay. 

The problem you experienced was due to ill-formed logic in the check for input errors. The regex actually successfully captured your message. Nonetheless, I expanded the regex because your concern about difficulty inputting messages in the wrong place is valid.

I fixed the error logic in https://github.com/woocommerce/woocommerce/pull/38510/commits/d0a08c0f54f013dc1e00577e81c96945063e66f9

I addressed the typescript optional parameters comment in https://github.com/woocommerce/woocommerce/pull/38510/commits/05e21d213da17578e993da01af44bac56eb0cae4

And I expanded the regex in https://github.com/woocommerce/woocommerce/pull/38510/commits/a2169092ce4ae9f449dc193cd89968e95c651783

Here are examples of inputs that are successfully captured by the regex. You can have [a play around with it here if you'd like](https://regex101.com/r/3VXwqv/1).

```
#### Message <!-- Add a changelog message here -->

This is a good message.
```
Message on same line as heading
```
#### Message <!-- Add a changelog message here --> This is a good message.
```
Comment deleted
```
#### Message

This is a good message.
```
No spaces after heading and comment deleted
```
#### Message
This is a good message.
```
No new line after heading
```
#### Message <!-- Add a changelog message here -->
This is a good message.
```

psealock committed:
lockfile

samueljseay reviewed (approved):
Apologies for the delay in re-review. After testing again this works well for me. Thanks for adding this I think it will be a great QOL improvement.

psealock committed:
lockfile
